### PR TITLE
ipopt: use url_for_version to handle different urls for 3.13.4 onwards

### LINKED ISSUE #30183: ipopt package can't construct proper url for fetch?
### Steps to reproduce

Something is wrong with the way `ipopt` package gets the URL for its versions.

```
$> spack -d stage ipopt@3.14.0
...
==> [2022-04-20-15:19:59.610933] Creating stage lock spack-stage-ipopt-3.14.0-4v4ov6bdt5nj2atwg7npsikri72frr5u
==> [2022-04-20-15:19:59.611258] Reading config file /spack/etc/spack/defaults/mirrors.yaml
==> [2022-04-20-15:19:59.614084] Checking existence of https://mirror.spack.io/_source-cache/archive/9b/9bed72a5456ef37f1b95746c932986e6664eb70b983d4fab61cf8aa811facdf1.tgz
==> [2022-04-20-15:19:59.724619] FailedDownloadError: Failed to fetch file from URL: https://mirror.spack.io/_source-cache/archive/9b/9bed72a5456ef37f1b95746c932986e6664eb70b983d4fab61cf8aa811facdf1.tgz
    Urllib fetch failed to verify url                      https://mirror.spack.io/_source-cache/archive/9b/9bed72a5456ef37f1b95746c932986e6664eb70b983d4fab61cf8aa811facdf1.tgz
 with error Download failed: HTTP Error 404: Not Found
==> [2022-04-20-15:19:59.725252] Checking existence of https://mirror.spack.io/ipopt/ipopt-3.14.0.tgz
==> [2022-04-20-15:19:59.837050] FailedDownloadError: Failed to fetch file from URL: https://mirror.spack.io/ipopt/ipopt-3.14.0.tgz
    Urllib fetch failed to verify url                      https://mirror.spack.io/ipopt/ipopt-3.14.0.tgz
 with error Download failed: HTTP Error 404: Not Found
==> [2022-04-20-15:19:59.837721] Checking existence of https://www.coin-or.org/download/source/Ipopt/Ipopt-3.14.0.tgz
==> [2022-04-20-15:20:02.146266] FailedDownloadError: Failed to fetch file from URL: https://www.coin-or.org/download/source/Ipopt/Ipopt-3.14.0.tgz
    Urllib fetch failed to verify url                      https://www.coin-or.org/download/source/Ipopt/Ipopt-3.14.0.tgz
 with error Download failed: HTTP Error 404: Not Found
==> [2022-04-20-15:20:02.146883] Fetching from https://mirror.spack.io/_source-cache/archive/9b/9bed72a5456ef37f1b95746c932986e6664eb70b983d4fab61cf8aa811facdf1.tgz failed.
==> [2022-04-20-15:20:02.147073] Fetching from https://mirror.spack.io/ipopt/ipopt-3.14.0.tgz failed.
==> [2022-04-20-15:20:02.147147] Fetching from https://www.coin-or.org/download/source/Ipopt/Ipopt-3.14.0.tgz failed.
==> [2022-04-20-15:20:02.147493] FetchError: All fetchers failed for spack-stage-ipopt-3.14.0-4v4ov6bdt5nj2atwg7npsikri72frr5u
==> [2022-04-20-15:20:02.147690] Error: All fetchers failed for spack-stage-ipopt-3.14.0-4v4ov6bdt5nj2atwg7npsikri72frr5u
Traceback (most recent call last):
  File "/spack/lib/spack/spack/main.py", line 915, in main
    return _main(argv)
  File "/spack/lib/spack/spack/main.py", line 870, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/spack/lib/spack/spack/main.py", line 898, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/spack/lib/spack/spack/main.py", line 550, in _invoke_command
    return_val = command(parser, args)
  File "/spack/lib/spack/spack/cmd/stage.py", line 66, in stage
    package.do_stage()
  File "/spack/lib/spack/spack/package.py", line 1552, in do_stage
    self.do_fetch(mirror_only)
  File "/spack/lib/spack/spack/package.py", line 1531, in do_fetch
    self.stage.fetch(mirror_only, err_msg=err_msg)
  File "/spack/lib/spack/spack/util/pattern.py", line 19, in __call__
    for item in self.container]
  File "/spack/lib/spack/spack/util/pattern.py", line 19, in <listcomp>
    for item in self.container]
  File "/spack/lib/spack/spack/stage.py", line 510, in fetch
    raise fs.FetchError(err_msg or default_msg, None)
spack.fetch_strategy.FetchError: All fetchers failed for spack-stage-ipopt-3.14.0-4v4ov6bdt5nj2atwg7npsikri72frr5u
```

```
$> spack -d stage ipopt@3.13.4
...
==> [2022-04-20-15:20:56.229004] Creating stage lock spack-stage-ipopt-3.13.4-cxkirdlsceqr7x3aikwvozfdd65a6lag
==> [2022-04-20-15:20:56.229298] Reading config file /spack/etc/spack/defaults/mirrors.yaml
==> [2022-04-20-15:20:56.231788] Checking existence of https://mirror.spack.io/_source-cache/archive/1f/1fdd0f8ea637856d66b1ebdd7d52ad1b8b8c1142d1a4ce0976b200ab280e5683.tgz
==> [2022-04-20-15:20:56.307167] FailedDownloadError: Failed to fetch file from URL: https://mirror.spack.io/_source-cache/archive/1f/1fdd0f8ea637856d66b1ebdd7d52ad1b8b8c1142d1a4ce0976b200ab280e5683.tgz
    Urllib fetch failed to verify url                      https://mirror.spack.io/_source-cache/archive/1f/1fdd0f8ea637856d66b1ebdd7d52ad1b8b8c1142d1a4ce0976b200ab280e5683.tgz
 with error Download failed: HTTP Error 404: Not Found
==> [2022-04-20-15:20:56.307829] Checking existence of https://mirror.spack.io/ipopt/ipopt-3.13.4.tgz
==> [2022-04-20-15:20:56.391054] FailedDownloadError: Failed to fetch file from URL: https://mirror.spack.io/ipopt/ipopt-3.13.4.tgz
    Urllib fetch failed to verify url                      https://mirror.spack.io/ipopt/ipopt-3.13.4.tgz
 with error Download failed: HTTP Error 404: Not Found
==> [2022-04-20-15:20:56.391609] Checking existence of https://www.coin-or.org/download/source/Ipopt/Ipopt-3.13.4.tgz
==> [2022-04-20-15:20:57.877299] FailedDownloadError: Failed to fetch file from URL: https://www.coin-or.org/download/source/Ipopt/Ipopt-3.13.4.tgz
    Urllib fetch failed to verify url                      https://www.coin-or.org/download/source/Ipopt/Ipopt-3.13.4.tgz
 with error Download failed: HTTP Error 404: Not Found
==> [2022-04-20-15:20:57.877919] Fetching from https://mirror.spack.io/_source-cache/archive/1f/1fdd0f8ea637856d66b1ebdd7d52ad1b8b8c1142d1a4ce0976b200ab280e5683.tgz failed.
==> [2022-04-20-15:20:57.878115] Fetching from https://mirror.spack.io/ipopt/ipopt-3.13.4.tgz failed.
==> [2022-04-20-15:20:57.878204] Fetching from https://www.coin-or.org/download/source/Ipopt/Ipopt-3.13.4.tgz failed.
==> [2022-04-20-15:20:57.878589] FetchError: All fetchers failed for spack-stage-ipopt-3.13.4-cxkirdlsceqr7x3aikwvozfdd65a6lag
==> [2022-04-20-15:20:57.878867] Error: All fetchers failed for spack-stage-ipopt-3.13.4-cxkirdlsceqr7x3aikwvozfdd65a6lag
Traceback (most recent call last):
  File "/spack/lib/spack/spack/main.py", line 915, in main
    return _main(argv)
  File "/spack/lib/spack/spack/main.py", line 870, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/spack/lib/spack/spack/main.py", line 898, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/spack/lib/spack/spack/main.py", line 550, in _invoke_command
    return_val = command(parser, args)
  File "/spack/lib/spack/spack/cmd/stage.py", line 66, in stage
    package.do_stage()
  File "/spack/lib/spack/spack/package.py", line 1552, in do_stage
    self.do_fetch(mirror_only)
  File "/spack/lib/spack/spack/package.py", line 1531, in do_fetch
    self.stage.fetch(mirror_only, err_msg=err_msg)
  File "/spack/lib/spack/spack/util/pattern.py", line 19, in __call__
    for item in self.container]
  File "/spack/lib/spack/spack/util/pattern.py", line 19, in <listcomp>
    for item in self.container]
  File "/spack/lib/spack/spack/stage.py", line 510, in fetch
    raise fs.FetchError(err_msg or default_msg, None)
spack.fetch_strategy.FetchError: All fetchers failed for spack-stage-ipopt-3.13.4-cxkirdlsceqr7x3aikwvozfdd65a6lag
```

### Error message

_No response_

### Information on your system

* **Spack:** 0.18.0.dev0 (c664a31f7b1016d060682ff458c4bae8067cff59)
* **Python:** 3.7.7
* **Platform:** darwin-monterey-cannonlake
* **Concretizer:** clingo

@felliu @ashermancinelli @adrienbernede 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by eugeneswalker**
Fixes https://github.com/spack/spack/issues/30183 using `url_for_version` per suggestion of ipopt power users (see linked issue)

Adds versions:
* 3.14.1
* 3.14.2
* 3.14.3
* 3.14.4

@ashermancinelli @felliu

### CODE REVIEW

**Review by ashermancinelli — APPROVED** [review]

**Comment by goxberry** [comment]
If this approach works, I think using `url_for_version` is fine for now, and I support merging this PR.

Longer-term, I think a more sustainable strategy will be to use the URLs associated with version tags from [Ipopt's GitHub repo](https://github.com/coin-or/Ipopt). The [COIN-OR projects web page](https://www.coin-or.org/projects/) now directs users to Ipopt's GitHub repo anyway, and I don't see any links on COIN-OR's website that lead to the old Trac-based Ipopt project page and to the source links in the current upstream version of `ipopt/package.py`.

I have to submit some Spack PRs today anyway, so I'd be happy to make further URL changes one of those PRs. I don't want to hold up this PR.

**Review by goxberry — APPROVED** [review]

**Review by wspear — APPROVED** [review]
Looks good to me.

**Comment by goxberry** [comment]
I wrote the initial version of the Ipopt Spack package, so feel free to add me as a maintainer.

**Pull request merged**
